test/src/test/java/hudson/slaves/JNLPLauncherRealTest.java

```diff
@@ -49,62 +47,66 @@
         /* Since RealJenkinsRuleInit.jpi will load detached plugins, to reproduce a failure use:
         FileUtils.touch(new File(rr.getHome(), "plugins/instance-identity.jpi.disabled"));
         */
-        rr.then(new ConnectStep(false));
-    }
-
-    private static class ConnectStep implements RealJenkinsRule.Step {
-
-      private final boolean webSocket;
-
-      private ConnectStep(boolean webSocket) {
-        this.webSocket = webSocket;
-      }
-
-      @Override
-      public void run(JenkinsRule r) throws Throwable {
-        InboundAgentRule inboundAgents = new InboundAgentRule(); // cannot use @Rule since it would not be accessible from the controller JVM
-        inboundAgents.apply(new Statement() {
-            @Override public void evaluate() throws Throwable {
-                for (PluginWrapper plugin : r.jenkins.pluginManager.getPlugins()) {
-                    System.err.println(plugin + " active=" + plugin.isActive() + " enabled=" + plugin.isEnabled());
-                }
-                assertThat(ExtensionList.lookupSingleton(JNLPLauncher.DescriptorImpl.class).doCheckWebSocket(webSocket, null).kind, is(FormValidation.Kind.OK));
-                InboundAgentRule.Options.Builder builder = InboundAgentRule.Options.newBuilder().name("static").secret();
-                if (webSocket) {
-                    builder.webSocket();
-                }
-                Slave agent = inboundAgents.createAgent(r, builder.build());
-                r.waitOnline(agent);
-                try {
-                    FreeStyleProject p = r.createFreeStyleProject();
-                    p.setAssignedNode(agent);
-                    FreeStyleBuild b = r.buildAndAssertSuccess(p);
-                    if (webSocket) {
-                        assertThat(agent.toComputer().getSystemProperties().get("java.class.path"), is(new File(r.jenkins.root, "agent.jar").getAbsolutePath()));
-                    }
-                    System.err.println(JenkinsRule.getLog(b));
-                } finally {
-                    inboundAgents.stop(r, agent.getNodeName());
-                }
-            }
-        }, Description.EMPTY).evaluate();
-
-      }
+        then(false);
     }
 
     /**
      * Simplified version of {@link WebSocketAgentsTest#smokes} just checking Jetty/Winstone.
      */
     @Issue("JENKINS-68933")
     @Test public void webSocket() throws Throwable {
-        rr.then(new ConnectStep(true));
+        then(true);
     }
 
+    private void then(boolean websocket) throws Throwable {
+        try {
+            rr.startJenkins();
+            InboundAgentRule.Options.Builder options = InboundAgentRule.Options.newBuilder().name(STATIC_AGENT_NAME);
+            if (websocket) {
+                options = options.webSocket();
+            }
+            iar.createAgent(rr, options.build());
+            rr.runRemotely(new RunJobStep(STATIC_AGENT_NAME, websocket));
+        } finally {
+            iar.stop(rr, STATIC_AGENT_NAME);
```

Comment:
Is this actually necessary, or can the `try` block be deleted and let the `@Rule` do its work?

Reply:
https://github.com/jenkinsci/jenkins-test-harness/pull/501#issuecomment-1284448230

Reply:
Thanks, I knew there was a discussion somewhere. Anyway, seems like once the apparent bug in production code is fixed, we _could_ delete these and let the rule cleanup run. So would be good to leave in a comment to that effect.

Reply:
https://github.com/jenkinsci/jenkins-test-harness/pull/501#issuecomment-1284480345